bindings/python/pyproject.toml

```
@@ -17,7 +17,7 @@ requires = [
     "setuptools>=42",
     "scikit-build",
     "cmake>=3.21",      # Keep in-sync with `CMakeLists.txt`
-    "numpy>=1.10.0, <2",    # Keep in-sync with `setup.py`
+    "numpy>=2.2, <3",    # Keep in-sync with `setup.py`
```

I think we should support 2.0 and 2.1 too. The deprecation schedule [suggested](https://numpy.org/neps/nep-0029-deprecation_policy.html) by NumPy shows them available for a long time:
https://scientific-python.org/specs/spec-0000/#2026---quarter-2
```suggestion
    "numpy>=2, <3",    # Keep in-sync with `setup.py`
```

---

i think that for build requirements we should pin exact versions instead of >= so we would have reproducible results 

While for install_requires we can use ranges. 

As for numpy specifically we should build with 2.0 and put run dependency as >=1.16  The way how numpy went with 2.0 migration is that they support this configuration.

---

As far as I understand, a reproducible build requires exact pinning for both install and build requirements. I think `pyproject.toml` and `setup.py` are not the best tools for exact pinning. I recommend looking at `uv` if we want to use exact pinning:
https://docs.astral.sh/uv/concepts/projects/layout/#the-lockfile

---

yes, uv would work fine.  but some changes are still required for numpy - building with 2.0 and allowing run dependency for 1.16 +